engine/utils/Colour.h

```diff
@@ -9,39 +9,203 @@
+        a {static_cast<T>(other.a)}
+    {}
 
-    static const Colour Pink;
+    /**
+     * A constructor which takes in the float colour type
+     * @param other the 8-bit unsigned integer based colour struct
+     */
+    inline constexpr Colour(const IColour& other) :
+        r {static_cast<T>(other.r)},
+        g {static_cast<T>(other.g)},
+        b {static_cast<T>(other.b)},
+        a {static_cast<T>(other.a)}
+    {}
 
-    // 'Structors
+    /**
+     * The equality operator for the float-based colour struct
+     * @param other the colour to test equality with
+     * @return a boolean specifying if the two types are equal
+     */
+    inline constexpr bool operator==(const FColour& other) const
+    {
+        return r == other.r && g == other.g && b == other.b && a == other.a;
+    }
 
-    Colour() : Colour(0, 0, 0, 0) {}
+    /**
+     * The equality operator for the 8-bit unsigned integer-based colour struct
+     * @param other the colour to test equality with
+     * @return a boolean specifying if the two types are equal
+     */
+    inline constexpr bool operator==(const IColour& other) const
+    {
+        return r == other.r && g == other.g && b == other.b && a == other.a;
+    }
 
-    Colour(int r, int g, int b) : r(r), g(g), b(b), a(255) {}
+    /**
+     * The inequality operator for the float-based colour struct
+     * @param other the colour to test inequality with
+     * @return a boolean specifying if the two types are not equal
+     */
+    inline constexpr bool operator!=(const FColour& other) const
+    {
+        return r == other.r || g == other.g || b == other.b || a == other.a;
+    }
 
-    Colour(int r, int g, int b, int a) : r(r), g(g), b(b), a(a) {}
+    /**
+     * The inequality operator for the 8-bit unsigned integer-based colour struct
+     * @param other the colour to test inequality with
+     * @return a boolean specifying if the two types are not equal
+     */
+    inline constexpr bool operator!=(const IColour& other) const
+    {
+        return r == other.r || g == other.g || b == other.b || a == other.a;
+    }
 
     // Public members
 
-    int r;
-    int g;
-    int b;
-    int a;
+    T r {0};
+    T g {0};
+    T b {0};
+    T a {0};
 };
+
+// FColour constants
+
+inline constexpr FColour FWhite = {1.f, 1.f, 1.f, 1.f};
+inline constexpr FColour FBlack = {0.f, 0.f, 0.f, 1.f};
+inline constexpr FColour FRed = {1.f, 0.f, 0.f, 1.f};
+inline constexpr FColour FGreen = {0.f, 1.f, 0.f, 1.f};
+inline constexpr FColour FBlue = {0.f, 0.f, 1.f, 1.f};
+inline constexpr FColour FPink = {1.f, 0.427450980392, 0.760784313725, 1.f};
+
+// IColour constants
+
+inline constexpr IColour IWhite = {255, 255, 255, 255};
+inline constexpr IColour IBlack = {0, 0, 0, 255};
+inline constexpr IColour IRed = {255, 0, 0, 255};
+inline constexpr IColour IGreen = {0, 255, 0, 255};
+inline constexpr IColour IBlue = {0, 0, 255, 255};
+inline constexpr IColour IPink = {255, 109, 194, 255};
```

Comment:
Move these inside their related classes and rename them without the prefixed e.g. `IColour::Pink`

Reply:
It took some playing around, but I got it working in the end.